Refactor antag rule code

Rainfey opened the pull request:
## About the PR
Refactor antag code

## Progress
Initial Pass
- [X] Thief
- [X] Traitor
- [X] Revolution
- [X] Zombies
- [X] NukeOps

Initial Testing
- [X] Thief
- [X] Traitor
- [X] Revolution
- [X] Zombies
- [X] NukeOps

Initial testing is complete, however i cant test these changes at full scale.

## Why / Balance
Antag rules duplicated the same code (roughly) for:
- Sending 'not enough players' warnings to admins
- Checking if there are enough players
- Selecting eligible players
- Choosing antags from eligible players
- Sending briefings to antags

Antags rules were not aware of other rules, resulting in multiple antag roles per player (see Discord comments regarding Thief/HeadRev's)
As antag roles are balanced around themselves, allowing multiple antatg's (per player) upsets this balance

## Technical details
- Remove duplicated code from antag rules, move to AntagSelectionSystem
- Simplify antag code, using inbuilt functions where possible
- Options to ignore job restrictions, multiple antag restrictions and user preferences when choosing antags
- Excluding invalid choices (ie disconnected players) handled centrally
- Added task scheduler to GameRuleSystem to handle things like InitialInfected delays and round end checks
- Nukeops code cleaned up a little, unfilled nukie roles will be spawned as ghost roles (was a TODO).
- Added new CVar 'nukeops.spawn_ghost_roles' - Controls if unfufilled nukie roles are skipped or spawned as ghost roles (aware that ghost roles may give away the presence of nukies to any dead players or observers)

## Media
- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**
:cl:
- fix: Disallow multiple antag roles per player

asperger-sind commented:
Apologies if this is included but I did not see this.
Any chance this PR could also make it so antag roles (e.g. traitor, thief) are prioritized over jobs that disallow antag (e.g. secoff, command), so that if UristMcSecMain, who we assume has secoff on high and medical doctor on low, will be heavily biased towards rolling medical doctor as it is antag eligible if they roll an antagonist role.

As a possible concern this could theoretically lead to metagaming whether or not someone is an antagonist based on observations from multiple rounds, but I believe the benefits heavily outweigh the negatives here.

Not sure about implementation, but this might require restitching antagonist and job selection code so that antagonists like traitor are selected roundstart but the player is only informed that they are an antagonist later on to prevent antag rolling, unless the PR that implemented delayed traitor activation accounted for something like this.

edit: i can't read, line 3 of technical details implies this is possible but i already wrote this so aaaa

Rainfey commented:
By default all antag roles will skip Heads, Security and Training jobs.
Though that will be configurable with these changes, ie i dont see why the Medical Intern couldnt be a zombie.

As for delaying antag selection, some rules do this already (though its messy) - such as Traitors (5 minute delay) and Zombies (~15 minute delay)

I think i will leave it up to a future PR to add a standard method of delaying the rule start / antag selection.

BYONDFuckery commented:
I kinda DO want headrev thieves. I played that once and it was really fun.

deltanedas commented on Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:
```diff
@@ -602,145 +614,67 @@ private void OnPlayersSpawning(RulePlayerSpawningEvent ev)
-            var numNukies = MathHelper.Clamp(_playerManager.PlayerCount / playersPerOperative, 1, maxOperatives);
+            //Select Nukies
+            //Select Commander, priority : commanderEligible, agentEligible, operativeEligible, all players
+            var selectedCommander = _antagSelection.ChooseAntags(new List<ICommonSession>[] { commanderEligible, agentEligible, operativeEligible, ev.PlayerPool }, 1).FirstOrDefault();
```
ChooseAntags could be refactored a bit to take the number first then variadic args so you dont need to write the new List<ICommonSession>[] bit every time

Rainfey replied:
Completely forgot params existed

deltanedas commented on Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:
```diff
@@ -602,145 +614,67 @@ private void OnPlayersSpawning(RulePlayerSpawningEvent ev)
+            //Create the team!
+            //Provide a player session if possible, otherwise spawn them as ghost roles
+            var operatives = new List<NukieSpawn>();
+            operatives.Add(new NukieSpawn(selectedCommander ?? null, NukieType.Commander));
```
?? null :trollface:

Rainfey replied:
shush, i need more sleep clearly.

TheShuEd commented:
https://discord.com/channels/310555209753690112/1191584454372757534/1193093412325441586

I described my opinion here.  I think overlapping antagonists is a good idea, but it's worth separating the categories of major and minor antagonists.  eliminate the overlap of major roles, but allow additional minor roles.

deltanedas commented:
of course it should always be an option. might want thief traitor but not turf tagger rev

Rainfey commented:
Perhaps, personally im of the mind that there are many people who want to play an antag role - but only so many antag slots per round.

It seems unfair to (by random chance) have a single player take multiple of those slots up.
Then theres any attempt at balancing antags, imagine (random example) abductor headrev.

Rainfey commented:
Switched traitor back to allowing multiple antags.
I realise that the rules are processed alphabetically, which means you can have HeadRev/Thief or Nukie/Thief - but not Traitor/Thief (as traitor checks for duplicate antags and runs after thief)

Do we need more specific logic for this?
ie a whitelist of acceptable antag pairings

VasilisThePikachu commented:
Question: A big gripe i had with the antag system is that it choses antags anyway in cases where no one has it checked (Example: All the players with syndicate ticked managed to roll a head job. Traitor game rule has already started and can no longer be canceled. No one else has traitor ticked. Pick them anyway at random despite their preferences.)

Unless i missed something i dont see this logic.

Rainfey commented:
With the current implementation yes:
The AntagSelectionSystem will pick from players with the role enabled - if there is nobody with that role enabled then fallback to everyone who's job doesnt prohibit it (ie no heads, sec or training roles).

Nukies starts before roles are selected so it just picks in priority of Commander > Agent > Operatives > Everyone.
Zombies will pick from role enabled players first, then fill the shortfall from everyone else (ignoring job restrictions)

The changes included in this PR prevent this from happening outside of Zombies and Nukies.
Traitors have late join mechanics, rev's might need looking at as numbers do matter for them (late join head revs?).

I believe the logic behind this is that player preferences are not checked till after the round has started - its better to pick players who didnt want the role, than to effectively become an extended round with no antags.

VasilisThePikachu commented:
> its better to pick players who didnt want the role, than to effectively become an extended round with no antags.

I agree here, but we are doing that because we got no other choice.

With the current logic, an antag banned person or someone who litterly just joined and does not have antag unlocked can get it since they won the rng. Which is bad, even worse for mrp where a player may like to play a character who is all about serving NT and they get rolled antag. It sucks for them 

My implementation idea is that in these cases. The traitor players would be attempted to be picked before the round even begins. After roles are picked where you can check if they can still be antag

Maybe Its a terrible idea maybe it's the best. I'm just idea dumping. No need to do it in this pr

Rainfey commented:
I can only think that fixing that specific problem requires every antag to have late join support.
If you cant start with the full compliment of antags, you either: 
start short and accept the difficulty reduction
start short and add more later
or start full and pick from non role-enabled players.

Rainfey commented:
That should be possible, in StartGamerule you can count all the role enabled sessions and abort.
Currently theres no standard method to do this, so you would have to duplicate the call to GetEligibleSessions and CalculateAntagCount.

The only sticking point is how i think the secret preset works (been a while since i checked):
Choose random gamerule
If that fails try traitor
If that fails do extended

So that will need to be reworked to try all gamemodes first.

AJCM-git commented on Content.Server/GameTicking/Rules/Components/NukeopsRuleComponent.cs:
```diff
@@ -233,3 +233,10 @@ public enum WinCondition : byte
     AllNukiesAlive
 }
+
+public enum NukieType
+{
+    Commander,
+    Agent,
+    Operative
+}
```
i really dislike this, this is information you should be able to get from the entity at any moment. At least move the enum to a datafield in `NukeopsRoleComponent` and set it up when the role is added or something or infer it in some other way

Rainfey replied:
This isnt about knowing who is which role.
Its used by the spawning system to designate a team composition - previously it was just 
'If index is 0 then commander, if index is 1 then agent, else operative'

NukieType is sent to the spawning method along with the players session - where it is used to choose the name, gear and set the PrototypeId on the mind's NukeopsRoleComponent

AJCM-git replied:
I mean thats kind of the issue isnt it? The best way to handle it is changing nukies to spawn the entities first and then assign the roles

AJCM-git replied:
I mean thats kind of the problem isnt it? The best way to handle things like this is spawning the nukies entities first and then assign the roles i think

Rainfey replied:
I'm not following what you mean.

This is the only method that uses NukieType (other than OnPlayersSpawning, which assigns it to specific sessions)
https://github.com/space-wizards/space-station-14/blob/32beffe2ef099f6a2305389154ec2867d9ec8e9a/Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs#L802-L828

Its never stored in a component or used after spawning is completed - just a clearer version of what already exists with it hardcoding specific indexes in the list of sessions to roles:
https://github.com/space-wizards/space-station-14/blob/8f1d67014cc3c261ffba31226c6d6f89b0ce3d3d/Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs#L951-L953
https://github.com/space-wizards/space-station-14/blob/8f1d67014cc3c261ffba31226c6d6f89b0ce3d3d/Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs#L874-L901

It also allows for SpawnOperatives to be order agnostic - ie it doesnt need to be ordered Commander>Agent>Operatives.

metalgearsloth replied:
Can't you just make the data encapsulating it a struct? So name + role + gear, then just make a separate one for each nukie type and not tie it to an enum / switch statement.

Plus that way it's reuseable for other stuff eventually, e.g. ninja spawning.

Rainfey replied:
Removed the enums.
There are now NukeopSpawnPreset's defined in the component that contain the spawn details.

AJCM-git commented on Content.Server/GameTicking/Rules/RevolutionaryRuleSystem.cs:
```diff
@@ -69,27 +63,101 @@ public override void Initialize()
+        }
+
+        if (mind != null && mind.OwnedEntity.HasValue)
+            _antagSelection.SendBriefing(mind.OwnedEntity.Value, Loc.GetString("rev-role-greeting"), Color.Red, new SoundPathSpecifier("/Audio/Ambience/Antag/headrev_start.ogg"));
```
Sound path specifiers shouldnt be made like this, move it to a datafield in the component

Rainfey replied:
I can see why this was done, there is no easy access to RevolutionaryRuleComponent in this method.
Should i convert the SoundSpecifier to a static field?
Otherwise i will need to either duplicate the SoundSpecifier on RevolutionaryComponent or query GameRules whenever someone is flashed.

AJCM-git replied:
yeah just repeat the field, not that big of a deal (tho i would assume the RevolutionaryComponent would be accessible whenever `RevolutionaryRuleComponent.HeadRevStartSound` is read, if so just move it from the rule. Havent checked thom could be wrong)

AJCM-git reviewed (changes requested):
Pushed a minor cleanup to your branch to speed the review process, check the commit to see if you disagree with any change.
Good pr, though some of the code feels off, did you use copilot or another AI to help you out?
Some reviews i have in general:
- Use ProtoId where necessary, there are some methods/fields that take a prototype ID but don't use this. (You can use `Index` instead of `TryIndex` when getting a ProtoId from a datafield since our serializer confirms it's existance)

- There seems to be some duplication on OnPlayerJobAssigned handlers though I didnt dive deep enough to confirm.

Rainfey commented on Content.Server/GameTicking/Rules/Components/GameRuleComponent.cs:
```diff
@@ -20,6 +20,8 @@ public sealed partial class GameRuleComponent : Component
     [DataField]
     public int MinPlayers;
+
+    public SortedList<TimeSpan, GameRuleTask> ScheduledTasks = new();
```
Fair enough, just seemed to be a commonality between game rules - usually for checking end conditions.
And thus was a candidate for making a standard method.

Rainfey replied:
Would it be acceptable to replace these calls with calls to Timer.Spawn and Timer.SpawnRepeating?

AJCM-git replied:
no, those are obsoleted

Rainfey commented on Content.Server/Antag/AntagSelectionSystem.cs:
```diff
@@ -44,177 +46,256 @@ public sealed class AntagSelectionSystem : GameRuleSystem<GameRuleComponent>
+    /// <param name="eligiblePlayers">List of eligible players</param>
+    /// <param name="count">How many to choose</param>
+    /// <returns>Up to the specified count of elements from the provided list</returns>
+    public List<T> ChooseAntags<T>(int count, List<T> eligiblePlayers)
```
Does not need to be a generic as most game rules use EntityUid, excluding Pirate and NukeOps, which use ICommonSession.
I could duplicate the method if that is preferred.

deltanedas replied:
probably better to make pirate and nukie rules to use mind ids instead of sessions

Rainfey replied:
Dont think i can, are minds not created when a session is bound to an entity?
Nukies run before any entities are spawned, and hijack the spawning process to force the sessions into specially spawned nuclear operative bodies.
Same for pirates

deltanedas replied:
oh true

AJCM-git replied:
> Does not need to be a generic as most game rules use EntityUid, excluding Pirate and NukeOps, which use ICommonSession. I could duplicate the method if that is preferred.

Well if they are ran before attaching entities thats a little cursed but add a ICommonSession version i guess

AJCM-git commented on Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:
```diff
@@ -607,145 +616,64 @@ private void OnPlayersSpawning(RulePlayerSpawningEvent ev)
-                {
-                    continue;
-                }
+            var nukiesToSelect = 6;//_antagSelection.CalculateAntagNumber(_playerManager.PlayerCount, nukeops.PlayersPerOperative, nukeops.MaxOps);
```
fix this

Rainfey replied:
The shame, forgot to remove it after testing.

Rainfey commented on Content.Server/Administration/Systems/AdminVerbSystem.Antags.cs:
All now take an EntityUid, since thats what the verbs provide directly.

Rainfey commented on Content.Server/Antag/AntagSelectionSystem.cs:
```diff
@@ -44,177 +46,256 @@ public sealed class AntagSelectionSystem : GameRuleSystem<GameRuleComponent>
         }
+
+        //Only allow humanoids
+        if (!HasComp<HumanoidAppearanceComponent>(session.AttachedEntity))
```
I made this an option argument on the selector, defaulting to false.

Rainfey commented on Content.Server/GameTicking/Rules/ThiefRuleSystem.cs:
```diff
@@ -195,13 +165,12 @@ private void OnGetBriefing(Entity<ThiefRoleComponent> thief, ref GetBriefingEven
     private string MakeBriefing(EntityUid thief)
```
Reverted to the original function, there was no real need to change it (micro optimisation bad)

Rainfey commented on Content.Server/Antag/AntagSelectionSystem.cs:
Moved IsGroupDead into RevolutionaryRuleSystem - it can be moved out should another system need it.
I think NukeOps has something similar, but i want to keep my touches light on that can of worms

GiveAntagBagGear moved to Content.Shared.Inventory.InventorySystem.Helpers as SpawnItemsOnEntity
Does a best attempt at putting the item/s in an entities bag, pockets, hand or nearby.

Rainfey commented:
No copilot, just the formatting tools included with VS. 
Though i did see some methods had been shifted around for no apparent reason - moved them back.
I converted a few easy fields to ProtoId, and moved them onto components - along with AntagSelectionSystem methods.

Rainfey commented:
As a solution to the multiple antag per player problem i mentioned in an earlier comment - specifically about it being alphabetical 
I have implemented an "ExclusiveAntag" attribute which should be applied to Antag Role Components.

Its purpose is to indicate to AntagSelectionSystem that (by default) - any entity possessing a role with this attribute should be ignored when selecting further antags.
This behaviour is overridable (see Thief) 

The default behaviour is thus:
If an entity has no antag roles, they are eligible
If an entity has a minor antag role, they are eligible
If an entity has a major antag role, they are not eligible

VasilisThePikachu commented:
Is this still being worked on by any chance?

Rainfey commented:
It is, bogged down with RL work at the moment so i find myself short on time

Rainfey commented:
Need to add a BecomesStationComponent onto the Nukie Outpost grid, 
along with adding the Nukie Outpost to the map list in Content.IntegrationTests.Tests.PostMapInitText

AJCM-git reviewed (approved):
fine to merge, tested as best as i could but there is a high chance something breaks in production

metalgearsloth commented:
Closer at least.
